### DISCUSSION TITLE
How it can be done in Mirth? I want to define a custom HL7 V2 (2.4) schema rather the default and to validate the incoming V2 messages

### DISCUSSION THREAD

**Discussion by xtrarima** [nextgenhealthcare/connect#5551] [General]
0

I want to define a custom HL7 V2 (2.4) schema rather the default and to validate the incoming V2 messages based on my custom schema defined.

i) The input message would be

MSH|^~&|Hey|Hey|Pharmacy|RE1267|20220717072035||ADT^A04^ADT_A05|20220720092046|P|2.4|||NE|NE|MYS|||| EVN||20220720092034 PID|1|MRN_X01|983103^^^^AD011||Mohamed^Fatimah^Noor^^NP003^^US004||19901021162034|FEMALE

ii) The user defined schema would be

<?xml version="1.0" encoding="utf-8"?>
<xsd:schema    xmlns:xsd="http://www.w3.org/2001/XMLSchema"    >
	
	
	<xsd:annotation>
		<xsd:appinfo source="urn:com.sun:encoder">
			<encoding xmlns="urn:com.sun:encoder" name="HL7 v2 Encoding" namespace="urn:com.sun:encoder-hl7-1.0" style="hl7encoder-1.0"/>
		</xsd:appinfo>
	</xsd:annotation>
	<xsd:element name="HL7Message" type="Message.Content"/>
	
	<xsd:complexType name="Message.Content">
		<xsd:sequence>
			<xsd:element name="MSH" type="MSH.Content" minOccurs="1" maxOccurs="1"/>
		</xsd:sequence>
	</xsd:complexType>
	<xsd:complexType name="MSH.Content">
		<xsd:sequence>
			<xsd:element name="MSH.1" type="xsd:string" />
			<xsd:element name="MSH.2" type="xsd:string"/>
			<xsd:element name="MSH.3" minOccurs="0">
				<xsd:complexType>
					<xsd:sequence>
						<xsd:element name="MSH.3.1" type="xsd:string" />
					</xsd:sequence>
				</xsd:complexType>
			</xsd:element>
		</xsd:sequence>
	</xsd:complexType>
</xsd:schema>

ut, I'm not sure how it can be achieved in Mirth.

My question is that does Mirth allow to define custom HL7 message and so parse the input?

If it is possible then could anyone explain how it can be done pls ?

Note: I've attached a sample xsd (pls consider it as sample alone) for reference.

**Comment by tom08zehn** [comment]
First of all, I guess the example HL7 is anonymized and only contains faked/random PHI...

I'd say this is not necessarily a Mirth specific question. Depends on what/how you want to achieve it.

AFAIK Mirth's built-in HL7v2 / ER7 parser "simply" grabs the characters from the MSH.1 and MSH.2 fields and splits the HL7v2-string into chunks. I'm not sure what happens when you enable "strict" parsing in the connector settings....

So as long as the HL7 is well-formed, it goes through and Mirth converts it into the internal XML format (serialized `<HL7Message>`).

In a next step, you could validate the XML against your schema.

See here how you could do that in Java: https://letmegooglethat.com/?q=java+xml+schema+validation
First result looks promising: https://www.digitalocean.com/community/tutorials/how-to-validate-xml-against-xsd-in-java

The question then is if you want to create your own Java library and load it into Mirth or if you simply want to do it in Mirth (Rhino) JavaScript. If JavaScript, then I recommend wrapping your code into a Code Template Library function (e.g. `xml_validate(xml, xsd)`) and call it in your preprocessor\* or source transformer script like so:

\* Note: In the preprocessor scope, the inbound message is still ER7 format (raw string) and you need to convert it manually to XML then. No big deal but I'd let Mirth do the work here.

Example for the source transformer script:

```js
const xsd = FileUtil.read('/path/to/schema.xsd');
const valid = xml_validate(msg, xsd); // returns true or false`

if(!valid) throw new Error('HL7 message not matching schema'); // stop message processing

// or if you want a silent/graceful stop
if(!valid) {
	logger.error('HL7 message not matching schema');
	destinationSet.removeAll();
	return;
}
```

Maybe you could also try JSON validation instead of XML... but not sure if this is built-in.

--

This could also be an interesting place to look at: https://github.com/nextgenhealthcare/connect/tree/master/server/src/com/mirth/connect/plugins/datatypes/edi

**Comment by tom08zehn** [comment]
Just found this, maybe helpful: https://forums.mirthproject.io/forum/mirth-connect/support/9318-working-solution-to-validate-a-xml-against-a-xsd

**Comment by xtrarima** [comment]
Thank you Tom for the reply. I must appreciate it. I understood the logic behind it before and did a try to parse it against an xsd. It was resulted in an error. But, the code snippet that used is different.
Let me try the new snippet provided. Anyway, I assume that Mirth does not support feature like how ensemble and other integration engine are allowing us to define custom schema and validate the input against it. In business, there are situations where we can't enable strict parser and we can't write validation code for all the segments/fields as well. it would be the case that I'm facing.

**Reply by tom08zehn** [reply]
Interesting... I never faced that requirement.

May I ask what other integration engine you mean and why or in what situations you need to validate a custom schema for HL7v2?

Sounds like it's something that you need quite often.

Also, what is the special thing in your posted schema?

**Comment by tom08zehn** [comment]
Just tried one of the examples that I've googled and works...

Following example has one XML document and two XSD schemas -> one of them is valid and one invalid - or the opposite: the XML is valid for just one of the XSDs... you know what I mean.

I was lazy and simply tried this in the deploy script of a new channel:

<details><summary>Full code here (click to expand)</summary>
<p>

```js
// https://www.baeldung.com/java-validate-xml-xsd

var xml = <?xml version="1.0" encoding="UTF-8" ?>
<individual>
    <name>Baeldung</name>
    <address>
        <zip>00001</zip>
        <city>New York</city>
    </address>
</individual>;

var xsdValid = <?xml version="1.0" encoding="UTF-8" ?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <xs:element name="individual">
        <xs:complexType>
            <xs:sequence>
                <xs:element name="name" type="xs:string" />
                <xs:element name="address">
                    <xs:complexType>
                        <xs:sequence>
                            <xs:element name="zip" type="xs:positiveInteger" />
                            <xs:element name="city" type="xs:string" />
                        </xs:sequence>
                    </xs:complexType>
                </xs:element>
            </xs:sequence>
        </xs:complexType>
    </xs:element>
</xs:schema>;

var xsdInvalid = <?xml version="1.0" encoding="UTF-8" ?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <xs:element name="individual">
        <xs:complexType>
            <xs:sequence>
                <xs:element name="name">
                    <xs:simpleType>
                        <xs:restriction base="xs:string">
                            <xs:maxLength value="5" />
                        </xs:restriction>
                    </xs:simpleType>
                </xs:element>
                <xs:element name="address">
                    <xs:complexType>
                        <xs:sequence>
                            <xs:element name="zip" type="xs:positiveInteger" />
                            <xs:element name="city" type="xs:string" />
                            <xs:element name="street" type="xs:string" />
                        </xs:sequence>
                    </xs:complexType>
                </xs:element>
            </xs:sequence>
        </xs:complexType>
    </xs:element>
</xs:schema>;

var xmlSource = new javax.xml.transform.stream.StreamSource(new java.io.StringReader(xml))

var factory = javax.xml.validation.SchemaFactory.newInstance(javax.xml.XMLConstants.W3C_XML_SCHEMA_NS_URI);

// test 1 (valid)
var xsdSource = new javax.xml.transform.stream.StreamSource(new java.io.StringReader(xsdValid));
var validator = factory.newSchema(xsdSource).newValidator();
try {
	validator.validate(xmlSource);
	logger.debug("valid");
} catch(e) {
	logger.debug("invalid");
}

// test 2 (invalid)
var xsdSource = new javax.xml.transform.stream.StreamSource(new java.io.StringReader(xsdInvalid));
var validator = factory.newSchema(xsdSource).newValidator();
try {
	validator.validate(xmlSource);
	logger.debug("valid");
} catch(e) {
	logger.debug("invalid");
}
```
</p>
</details>


Result:

```
DEBUG 2022-12-09 21:17:21.633 [Deploy JavaScript Task on xsd_validation (bbefa0ab-fa44-499f-9071-177626f21269) < pool-1-thread-7] xsd_validation:deploy: valid
DEBUG 2022-12-09 21:17:21.755 [Deploy JavaScript Task on xsd_validation (bbefa0ab-fa44-499f-9071-177626f21269) < pool-1-thread-7] xsd_validation:deploy: invalid
```

**Reply by tom08zehn** [reply]
@xtrarima Please let us know if this works for you. Maybe you also find the time to answer my previous questions.
